### PR TITLE
Semantic place using varied gazing angles based on receptacle heights

### PR DESCRIPTION
**Pull request by Achuthankrishna**
Added new files and made changes in gaze_env to read concept graph, extract height of the receptacle and finally adjust the gazing angle of the spot arm to enable waypoint estimation for higher receptacles.

Test results on various receptacles captured by Concept Graph in the apartment : https://docs.google.com/spreadsheets/d/159ZvnrZJgnAgJqd9qWFKWMEauIQv-kj9hoXPO5oTMDw/edit?gid=0#gid=0

### To use the place skill with waypoint estimation + CG for adjusting the arm joint angles use:
`skillmanager.place(place_target=receptacle_name, enable_waypoint_estimation=True)`
* `place_target=receptacle_name` should be a string representing the receptacle the robot should place on.
* `enable_waypoint_estimation= True` if you want to use the waypoint estimator.

The place skill will read from the concept graph json defined in `config.CONCEPT_GRAPH_FILE`  to determine the height of the receptacle and adjust the arm joint angels accordingly before running the waypoint estimator. If the `place_target` does not exist (or we choose to pass in None or an empty string), then we will set the arm to default joint angles.

### CODE REVIEW

File: spot_rl_experiments/configs/config.yaml [hunk 1]
```diff
@@ -19,6 +19,9 @@ WEIGHTS_TORCHSCRIPT:
   # Semantic place
   SEMANTIC_PLACE: "weights/semantic_place/sp12lhb6v10_12_ckpt.199.torchscript"
 
+  # Semantic place EE
+  SEMANTIC_PLACE_EE: "weights/semantic_place_ee/sp_jt_ckpt.16.torchscript"
```
Comment (joannetruong) [inline spot_rl_experiments/configs/config.yaml:23]: Semantic Place EE weights aren't relevant to this PR, so let's remove it for now. We can add this in the PR for adding EE skills

File: spot_rl_experiments/configs/config.yaml [hunk 2]
```diff
@@ -41,6 +44,9 @@ WEIGHTS:
...
```
Comment (joannetruong) [inline spot_rl_experiments/configs/config.yaml:48]: Remove from this PR

File: spot_rl_experiments/configs/config.yaml [hunk 3]
```diff
@@ -64,8 +70,6 @@ MAX_EPISODE_STEPS: 100
 # Nav env
 SUCCESS_DISTANCE: 0.3
 SUCCESS_ANGLE_DIST: 5
-SUCCESS_DISTANCE_FOR_DYNAMIC_YAW_NAV: 1.10
```
Comment (joannetruong) [inline spot_rl_experiments/configs/config.yaml:67]: Don't remove these config values

File: spot_rl_experiments/configs/config.yaml [hunk 4]
```diff
@@ -101,11 +105,14 @@ PLACE_ACTION_SPACE_LENGTH: 4
...
 
+# Semantic Place EE
+SEMANTIC_PLACE_EE_ACTION_SPACE_LENGTH: 10
+
 # Open Close Drawer env
 MAX_LIN_DIST_OPEN_CLOSE_DRAWER: 0.0
 MAX_ANG_DIST_OPEN_CLOSE_DRAWER: 3.73 # degrees
 OPEN_CLOSE_DRAWER_ACTION_SPACE_LENGTH: 8
-OPEM_CLOSE_DRAWER_DISTANCE_BETWEEN_EE_HANDLE: 0.75
+OPEM_CLOSE_DRAWER_DISTANCE_BETWEEN_EE_HANDLE: 0.3
```
Comment (joannetruong) [inline spot_rl_experiments/configs/config.yaml:115]: Not sure why the value changed here, let's keep the original value

File: spot_rl_experiments/configs/config.yaml [hunk 5]
```diff
@@ -101,11 +105,14 @@ PLACE_ACTION_SPACE_LENGTH: 4
 SEMANTIC_PLACE_ACTION_SPACE_LENGTH: 9
 SEMANTIC_PLACE_JOINT_BLACKLIST: [3]
 
+# Semantic Place EE
+SEMANTIC_PLACE_EE_ACTION_SPACE_LENGTH: 10
```
Comment (joannetruong) [inline spot_rl_experiments/configs/config.yaml:109]: Remove from this PR

File: spot_rl_experiments/configs/config.yaml [hunk 6]
```diff
@@ -135,8 +142,10 @@ INITIAL_ARM_JOINT_ANGLES_SEMANTIC_PLACE: [0, -180, 180, 0, 0, 90] # The initial
 INITIAL_ARM_JOINT_ANGLES_SEMANTIC_PLACE_LEFT_HAND: [0, -180, 180, 0, 0, -90] # The initial orientation of the arm for side grasping (gripper is in object's left hand side)
 INITIAL_ARM_JOINT_ANGLES_SEMANTIC_PLACE_TOP_DOWN: [-0.91, -92.87, 91.21, 0, 90.01, 0] # The initial orientation of the arm for top down grasping
 # The old gaze ready angle: [0, -170, 120, 0, 75, 0]
-GAZE_ARM_JOINT_ANGLES: [0, -160, 100, 0, 75, 0] #[0, -160, 100, 0, 90, 0]
-SEMANTIC_PLACE_ARM_JOINT_ANGLES: [0, -125, 80, 0, 85, 0]
```
Comment (joannetruong) [inline spot_rl_experiments/configs/config.yaml:139]: Is this value `SEMANTIC_PLACE_ARM_JOINT_ANGLES` being used anywhere else? Will removing this value break anything?

File: spot_rl_experiments/experiments/skill_test/test_sem_place.py [hunk 7]
```diff
@@ -14,25 +21,74 @@
...
         # at NYC
         place_target = "test_desk"
 
     spotskillmanager = SpotSkillManager(use_mobile_pick=False, use_semantic_place=True)
+    spot_pos1 = spotskillmanager.spot.get_arm_joint_positions(as_array=True)
+
+    # is_local = False
+    # if enable_estimation_before_place:
+    #     place_target = None
+    #     is_local = True
```
Comment (joannetruong) [inline spot_rl_experiments/experiments/skill_test/test_sem_place.py:35]: Remove this commented out code

File: spot_rl_experiments/experiments/skill_test/test_sem_place.py [hunk 8]
```diff
@@ -14,25 +21,74 @@
...
+        file_path = (
+            f"logs/semantic_place/{curr_date}/episode_sem_pl_run2_{episode_ctr}.json"
+        )
+        with open(file_path, "w") as file:
+            json.dump(episode_log, file, indent=4)
+            print(f"Saved log: {file_path}")
+        episode_ctr += 1
+        contnue = map_user_input_to_boolean("Do you want to do it again ? Y/N ")
+        # Return the arm to the original position
+        spot_pos = spotskillmanager.spot.get_ee_pos_in_body_frame()[0]
```
Comment (joannetruong) [inline spot_rl_experiments/experiments/skill_test/test_sem_place.py:90]: Why do we need to set the EE position instead of the initial joint angles?

File: spot_rl_experiments/experiments/skill_test/test_sem_place.py [hunk 9]
```diff
@@ -14,25 +21,74 @@
...
+    #     contnue = map_user_input_to_boolean("Do you want to do it again ? Y/N ")
     is_local = False
     if enable_estimation_before_place:
+        height_target = place_target
         place_target = None
         is_local = True
 
     # Start testing
     contnue = True
+    INITIAL_ARM_JOINT_ANGLES = [0, -180, 180, 90, 0, -90]
```
Comment (joannetruong) [inline spot_rl_experiments/experiments/skill_test/test_sem_place.py:51]: Is this being used somewhere? if not we can remove it

File: spot_rl_experiments/experiments/skill_test/test_sem_place.py [hunk 10]
```diff
@@ -14,25 +21,74 @@
...
     if in_fre_lab:
         # at FRE
-        place_target = "coffee_table"
+        place_target = "office chair"
     else:
         # at NYC
         place_target = "test_desk"
 
     spotskillmanager = SpotSkillManager(use_mobile_pick=False, use_semantic_place=True)
+    spot_pos1 = spotskillmanager.spot.get_arm_joint_positions(as_array=True)
```
Comment (joannetruong) [inline spot_rl_experiments/experiments/skill_test/test_sem_place.py:30]: Is this value being used anywhere? if not we can remove it

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 11]
```diff
@@ -1,15 +1,16 @@
...
-
-from typing import Any, Dict, List, Tuple
+from typing import Any, Dict, Tuple
 
 import magnum as mn
 import numpy as np
 import rospy
 from multimethod import multimethod
 from perception_and_utils.utils.generic_utils import conditional_print
+from spot_rl.envs.test_cfslam import calculate_height
```
Comment (joannetruong) [inline spot_rl_experiments/spot_rl/envs/skill_manager.py:13]: let's move the calculate_height stuff somewhere else, maybe within utils/utils.py? @jimmytyyang do you have any suggestions?

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 12]
```diff
@@ -231,8 +234,8 @@ def reset(self):
         # Reset the policies and environments via the controllers
         raise NotImplementedError
 
-    @multimethod
-    def nav(self, nav_target: str = None) -> Tuple[bool, str]:
+    @multimethod  # type: ignore
+    def nav(self, nav_target: str = None) -> Tuple[bool, str]:  # type: ignore
```
Comment (joannetruong) [inline spot_rl_experiments/spot_rl/envs/skill_manager.py:238]: can we remove this #type: ignore stuff?

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 13]
```diff
@@ -301,26 +304,6 @@ def nav(  # noqa
...
-    @multimethod  # type: ignore
-    def nav(self, x: float, y: float) -> Tuple[bool, str]:  # noqa
-        """
-        Perform the nav action on the navigation target with yaw specified as a metric location
-        Args:
-            x (float): x coordinate of the nav target (in meters) specified in the world frame
-            y (float): y coordinate of the nav target (in meters) specified in the world frame
-        Returns:
-            bool: True if navigation was successful, False otherwise
-            str: Message indicating the status of the navigation
```
Comment (joannetruong) [inline spot_rl_experiments/spot_rl/envs/skill_manager.py:313]: why is this removed?

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 14]
```diff
@@ -506,17 +488,26 @@ def place(self, place_target: str = None, ee_orientation_at_grasping: np.ndarray
                 is_local = True
         else:
             message = "No place target specified, estimating point through heuristic"
+            height = calculate_height(height_target)
+            if height < self.place_config.HEIGHT_THRESHOLD:
+                print("lower than 20 in")
```
Comment (joannetruong) [inline spot_rl_experiments/spot_rl/envs/skill_manager.py:493]: we can remove the prints here

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 15]
```diff
@@ -506,17 +488,26 @@ def place(self, place_target: str = None, ee_orientation_at_grasping: np.ndarray
...
+                self.angle = self.place_config.GAZE_ARM_JOINT_ANGLES_LOW_RECEPTACLES
+            else:
+                print("higher than 20 inc")
+                self.angle = self.place_config.GAZE_ARM_JOINT_ANGLES_HIGH_RECEPTACLES
+
             conditional_print(message=message, verbose=self.verbose)
             is_local = True
             # estimate waypoint
+            visualize = True
```
Comment (joannetruong) [inline spot_rl_experiments/spot_rl/envs/skill_manager.py:502]: remove this hardcoded value

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 16]
```diff
@@ -578,7 +569,7 @@ def contrainedplace(self, object_target: str = None, ee_orientation_at_grasping:
             object_target,
             proposition,
             self.spot,
-            self.pick_config.SEMANTIC_PLACE_ARM_JOINT_ANGLES,
+            self.pick_config.GAZE_ARM_JOINT_ANGLES,
```
Comment (joannetruong) [inline spot_rl_experiments/spot_rl/envs/skill_manager.py:572]: Does this function also use waypoint estimator? If so, we should make this the dynamic angle based on the target. @jimmytyyang for confirmation

File: spot_rl_experiments/spot_rl/envs/test_cfslam.py
Comment (joannetruong) [inline spot_rl_experiments/spot_rl/envs/test_cfslam.py:1]: we can move the calculate_height function somewhere else and delete this file

File: spot_rl_experiments/experiments/skill_test/test_sem_place.py [hunk 17]
```diff
@@ -14,25 +21,74 @@
...
+        with open(file_path, "w") as file:
+            json.dump(episode_log, file, indent=4)
+            print(f"Saved log: {file_path}")
+        episode_ctr += 1
+        contnue = map_user_input_to_boolean("Do you want to do it again ? Y/N ")
+        # Return the arm to the original position
+        spot_pos = spotskillmanager.spot.get_ee_pos_in_body_frame()[0]
+        spotskillmanager.spot.move_gripper_to_point(spot_pos, spot_ort)
 # The following is a helpful tip to debug the arm
 # We get Spot class
```
Comment (joannetruong) [inline spot_rl_experiments/experiments/skill_test/test_sem_place.py:93]: remove all unused code

File: spot_rl_experiments/spot_rl/envs/skill_manager.py (hunk 12 above)
Comment (Achuthankrishna) [reply spot_rl_experiments/spot_rl/envs/skill_manager.py:238]: Consulted with @jimmytyyang for passing pre-commit checks, and was asked to keep it since pre-commit checks were failing.

File: spot_rl_experiments/spot_rl/envs/skill_manager.py (hunk 11 above)
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/envs/skill_manager.py:13]: agree with what you said here!

File: spot_rl_experiments/spot_rl/envs/test_cfslam.py [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+# LICENSE file in the root directory of this source tree.
+# black: ignore-errors
+import json
+import os
+
+
+def calculate_height(object_tag):
+    # Iterate through each dictionary in the world_graph
+    os.chdir(
+        "/home/tusharsangam/Desktop/spot-sim2real/spot_rl_experiments/spot_rl/envs"
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/envs/test_cfslam.py:12]: use different way to get path, do not add user path here

File: spot_rl_experiments/spot_rl/envs/test_cfslam.py [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+    )
+    with open("cfslam_object_relations.json") as f:
+        world_graph = json.load(f)
+    print("THE OBJECT I GOT IS", object_tag)
+    for rel in world_graph:
+        # Iterate through all keys in the dictionary
+        for key, value in rel.items():
+            if isinstance(value, dict) and value.get("object_tag") == object_tag:
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/envs/test_cfslam.py:20]: does this mean that the name of `object_tag` needs to be exact? what if you cannot fine the name? we just throw an error?

File: spot_rl_experiments/spot_rl/envs/test_cfslam.py [hunk 20]
```diff
@@ -0,0 +1,31 @@
+# Copyright (c) Meta Platforms, Inc. and its affiliates.
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/envs/test_cfslam.py:1]: move this to util folder as Joanne said

File: spot_rl_experiments/spot_rl/envs/skill_manager.py (hunk 16 above)
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/envs/skill_manager.py:589]: yes, true, it uses waypoint esitmation as well. @Achuthankrishna could you update this as well?

File: spot_rl_experiments/experiments/skill_test/test_sem_place.py [hunk 21]
```diff
@@ -14,25 +21,60 @@
...
+        # Place Object and Close Gripper
         rospy.set_param("is_gripper_blocked", 0)
-        spotskillmanager.place(place_target, is_local=is_local, visualize=True)
-        contnue = map_user_input_to_boolean("Do you want to do it again ? Y/N ")
+        episode_log = {"actions": []}  # mypy: ignore-errors
+        spotskillmanager.spot.close_gripper()
+        input("waiting for user to get ready with camera")
+        if place_target is None:
+            spotskillmanager.place(
+                place_target, height_target, is_local=is_local, visualize=True
```
Comment (jimmytyyang) [inline spot_rl_experiments/experiments/skill_test/test_sem_place.py:55]: why there are two variables -- place_target and height_target? since based on the code, they are refering to the name of the object?

File: spot_rl_experiments/configs/config.yaml [hunk 22]
```diff
@@ -135,8 +137,10 @@ INITIAL_ARM_JOINT_ANGLES_SEMANTIC_PLACE: [0, -180, 180, 0, 0, 90] # The initial
...
+GAZE_ARM_JOINT_ANGLES: [0, -160, 100, 0, 75, 0]
+GAZE_ARM_JOINT_ANGLES_LOW_RECEPTACLES: [0, -160, 100, 0, 75, 0] #[0, -160, 100, 0, 90, 0]
+GAZE_ARM_JOINT_ANGLES_HIGH_RECEPTACLES: [0, -120, 60, 0, 88, 0] #[0, -160, 100, 0, 90, 0]
+HEIGHT_THRESHOLD : 0.8128
```
Comment (jimmytyyang) [inline spot_rl_experiments/configs/config.yaml:143]: why this value?

**Review by jimmytyyang — CHANGES_REQUESTED** [review]
looks nice! leave some comments, please let me know if there are something unclear

File: spot_rl_experiments/configs/config.yaml (hunk 22 above)
Comment (Achuthankrishna) [reply spot_rl_experiments/configs/config.yaml:143]: Hi Jimmy, myself and @joannetruong did teleop of the arm in the lab, and noted these values - the low receptacle being the old value (just the gripper now pointing down looking on the receptacle). The higher receptacle values were noted based on how the gripper camera viewed the table of 42 inch and higher.

Comment (Achuthankrishna) [reply spot_rl_experiments/configs/config.yaml:143]: For the Height_threshold, we decided to keep it as 32 inch table rn - as when we tested in the lab, objects above 32 inch in our setup [ie 42 in and 52in table setup] required higher gazing angle.

File: spot_rl_experiments/experiments/skill_test/test_sem_place.py (hunk 21 above)
Comment (Achuthankrishna) [reply spot_rl_experiments/experiments/skill_test/test_sem_place.py:55]: I added two variables for function testing. Basically in waypoint estimation, place_target would be "None", but I wanted to test the calculate_height function , so forced the place_target as none and created height_target to test out the function.

**Comment by joannetruong** [comment]
We should also modify any scripts that use `spotskillmanager.place(None)` and modify it so it follows the new convention right? i.e., spotskillmanager.place("", enable_waypoint_estimation=True)

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 23]
```diff
@@ -113,6 +114,7 @@ def __init__(
 
         # Process the meta parameters
         self._use_mobile_pick = use_mobile_pick
+        print("USE MOBILE PICK VARIABLE IS SET TO :", self._use_mobile_pick)
```
[the remark this answers is not in the document]
Comment (joannetruong) [reply spot_rl_experiments/spot_rl/envs/skill_manager.py:117]: reminder to remove print statement

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 24]
```diff
@@ -174,6 +176,7 @@ def __init_spot(
         )
 
         if place_config is None:
+            # TODO: overwrite the config
```
Comment (joannetruong) [inline spot_rl_experiments/spot_rl/envs/skill_manager.py:179]: remove TODO comment

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 25]
```diff
@@ -506,6 +509,16 @@ def place(self, place_target: str = None, ee_orientation_at_grasping: np.ndarray
                 is_local = True
         else:
             message = "No place target specified, estimating point through heuristic"
+            height = calculate_height(place_target)
```
Comment (joannetruong) [inline spot_rl_experiments/spot_rl/envs/skill_manager.py:512]: since we are using this block of code twice, we can make this into a function and call the function in both place and constrainedplace to avoid code duplication

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 26]
```diff
@@ -569,6 +583,17 @@ def contrainedplace(self, object_target: str = None, ee_orientation_at_grasping:
             "next-to",
         ], f"Place skill does not support proposition of {proposition}"
 
+        height = calculate_height(object_target)
+        if height < self.place_config.HEIGHT_THRESHOLD:
+            print("HERE")
+            self.arm_joint_angles = (
```
[the remark this answers is not in the document]
Comment (joannetruong) [reply spot_rl_experiments/spot_rl/envs/skill_manager.py:588]: reminder to resolve this comment. And remove print statement

File: spot_rl_experiments/spot_rl/utils/search_table_location.py [hunk 27]
```diff
@@ -410,6 +410,7 @@ def detect_place_point_by_pcd_method(
     print(f"Intel point {selected_point}, Gripper Point {selected_point_in_gripper}")
 
     img_with_bbox = None
+    visualize = True
```
Comment (joannetruong) [inline spot_rl_experiments/spot_rl/utils/search_table_location.py:413]: remove hardcoded visualize=True

File: spot_rl_experiments/experiments/skill_test/test_nav_pick_nav_place_skills.py [hunk 28]
```diff
@@ -19,7 +19,9 @@
             enable_force_control=True,
         )
         spotskillmanager.nav("kitchen_counter")
-        spotskillmanager.place(None, is_local=True, visualize=False)
+        spotskillmanager.place(
+            "", is_local=True, visualize=False, enable_waypoint_estimation=True
```
Comment (joannetruong) [inline spot_rl_experiments/experiments/skill_test/test_nav_pick_nav_place_skills.py:23]: we can still use spotskillmanager.place(None, ...) right? if so, let's change "" back to None

File: spot_rl_experiments/experiments/skill_test/test_sem_place.py [hunk 29]
```diff
@@ -14,38 +21,59 @@
...
+                place_target,
+                is_local=is_local,
+                visualize=False,
+                enable_waypoint_estimation=True,
+            )
+        else:
+            spotskillmanager.place(
+                place_target,
+                is_local=is_local,
+                visualize=True,
```
Comment (joannetruong) [inline spot_rl_experiments/experiments/skill_test/test_sem_place.py:60]: why is visualize=True for this one? 

we can also change the whole if/else block to:
```
if enable_estimation_before_place:
    place_target = None
spotskillmanager.place(
    place_target,
    is_local=enable_estimation_before_place,
    visualize=False,
     enable_waypoint_estimation=enable_estimation_before_place,
)
```

**Review by joannetruong — COMMENTED** [review]
let's make sure we resolve the comments so we can merge the PR soon!

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 30]
```diff
@@ -485,7 +489,7 @@ def place(self, place_target: str = None, ee_orientation_at_grasping: np.ndarray
             verbose=self.verbose,
         )
 
-        if place_target is not None:
+        if enable_waypoint_estimation is False:
```
Comment (jimmytyyang) [inline spot_rl_experiments/spot_rl/envs/skill_manager.py:492]: `if not enable_waypoint_estimation`

File: spot_rl_experiments/spot_rl/envs/skill_manager.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (jimmytyyang) [reply spot_rl_experiments/spot_rl/envs/skill_manager.py:588]: agree with what Joanne said here

**Review by jimmytyyang — CHANGES_REQUESTED** [review]

File: spot_rl_experiments/spot_rl/envs/skill_manager.py [hunk 31]
```diff
@@ -466,15 +467,28 @@ def semanticpick(
...
+        if height < self.place_config.HEIGHT_THRESHOLD:
+            self.arm_joint_angles = (
+                self.place_config.GAZE_ARM_JOINT_ANGLES_LOW_RECEPTACLES
+            )
+        else:
+            self.arm_joint_angles = (
+                self.place_config.GAZE_ARM_JOINT_ANGLES_HIGH_RECEPTACLES
+            )
+        return self.arm_joint_angles
+
```
Comment (joannetruong) [inline spot_rl_experiments/spot_rl/envs/skill_manager.py:481]:
```
def set_arm_joint_angles(self, place_target: str = None):
        height = calculate_height(place_target)
        if height < self.place_config.HEIGHT_THRESHOLD:
            return self.place_config.GAZE_ARM_JOINT_ANGLES_LOW_RECEPTACLES
        else:
            return self.place_config.GAZE_ARM_JOINT_ANGLES_HIGH_RECEPTACLES
```

**Review by joannetruong — CHANGES_REQUESTED** [review]
requesting a small change in the `set_arm_joint_angles` function, but should be ready to merge after this fix

**Review by joannetruong — APPROVED** [review]
looks good! thanks :)

**Review by jimmytyyang — APPROVED** [review]
LGTM!
